asapdiscovery-alchemy/asapdiscovery/alchemy/predict.py

```diff
@@ -0,0 +1,830 @@
+    add SMILES for each row.
+
+    Args:
+        dataframe: The pandas dataframe we should add the smiles to.
+        ligands: The list of ligands from which we should extract the smiles based on matching by name.
+    Returns:
+        The dataframe with added smiles column(s)
+    """
+    # get the experimental data for SMILES per compound name.
+    expt_smiles = {ligand.name: ligand.smiles for ligand in ligands}
+
+    # get the SMILES. Relative dataframe is a bit more involved than absolute.
+    if "labelA" in dataframe.columns:
+        smiles_a = []
+        smiles_b = []
+        for labelA, labelB in dataframe[["labelA", "labelB"]].values:
+            smiles_a.append(expt_smiles[labelA])
+            smiles_b.append(expt_smiles[labelB])
+        dataframe["SMILES_A"] = smiles_a
+        dataframe["SMILES_B"] = smiles_b
+
+    elif "label" in dataframe.columns:
+        smiles = [expt_smiles[label] for label in dataframe["label"].values]
+        dataframe["SMILES"] = smiles
+
+    return dataframe
+
+
+def extract_experimental_data(
+    reference_csv: str, assay_units: Literal["pIC50", "IC50"]
+) -> dict[str, tuple[unit.Quantity, unit.Quantity]]:
+    """
+    Extract the experimental data from the given csv file, this assumes the csv has been downloaded from cdd.
+    Where the molecule identifier is under column 'Molecule Name' and the experimental data is pIC50 / IC50
+    TODO make more general
+
+    Args:
+        reference_csv: The name of the csv file with the experimental data
+        assay_units: The assay units of 'pIC50' or 'IC50' that the experimental data is given in.
+
+    Returns:
+        A dictionary of molecule names and tuples of
+        experimental data and its associated uncertainty converted to Gibbs free energy in kcal/mol.
+    """
+    experimental_data = {}
+    assay_tags = [assay_units, assay_units + "_Mean"]
+    exp_data = pd.read_csv(reference_csv).fillna(0)
+
+    # work out the columns for the ref data and the uncertainty
+    assay_endpoint_tag, assay_endpoint_confidence_tag = None, None
+    for col in exp_data.columns:
+        for assay_tag in assay_tags:
+            if col.replace(" (µM)", "").endswith(assay_tag):
+                assay_endpoint_tag = col
+            elif col.replace(" (µM)", "").endswith(
+                f"{assay_tag} Standard Deviation (±)"
+            ):
+                assay_endpoint_confidence_tag = col
+    if assay_endpoint_tag is None:
+        raise RuntimeError(
+            f"Could not determine the assay tag from the provided units {assay_units}."
+        )
+    if assay_endpoint_confidence_tag is None:
+        warnings.warn(
+            f"Failed to detect Standard Deviation in experimental reference file {reference_csv}."
+        )
+
+    if assay_units == "pIC50":
+        converter = pic50_to_dg
+        units = unit.dimensionless
+    else:
+        converter = ki_to_dg
+        units = unit.molar
```

Comment:
@JenkeScheen should the units here then be micromolar if that's the standard used in CDD?

Reply:
yeah I think so